src/System.ComponentModel.EventBasedAsync/tests/AsyncOperationTests.cs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.ComponentModel;
+using System.Threading;
+using Xunit;
+
+namespace System.ComponentModel.EventBasedAsync
+{
+    public class AsyncOperationTests
+    {
+        private const int k_spinTimeoutSeconds = 20;
+
+        [Fact]
+        public static void Noop()
+        {
+            // Test that a simple AsyncOperation can be dispatched and completed via AsyncOperationManager
+            var operation = new TestAsyncOperation(op => {});
+            operation.Wait();
+            Assert.True(operation.Completed);
+            Assert.False(operation.Cancelled);
+            Assert.Null(operation.Exception);
+        }
+
+        [Fact]
+        public static void ThrowAfterAsyncComplete()
+        {
+            var operation = new TestAsyncOperation(op => {});
+            operation.Wait();
+
+            SendOrPostCallback noopCallback = state => { };
+            Assert.Throws<InvalidOperationException>(() => operation.AsyncOperation.Post(noopCallback, null));
+            Assert.Throws<InvalidOperationException>(() => operation.AsyncOperation.PostOperationCompleted(noopCallback, null));
+            Assert.Throws<InvalidOperationException>(() => operation.AsyncOperation.OperationCompleted());
+        }
+
+        [Fact(Skip="Bug 1092169")]
+        public static void ThrowAfterSynchronousComplete()
+        {
+            var operation = AsyncOperationManager.CreateOperation(null);
+            operation.OperationCompleted();
+
+            SendOrPostCallback noopCallback = state => {};
+            Assert.Throws<InvalidOperationException>(() => operation.Post(noopCallback, null));
+            Assert.Throws<InvalidOperationException>(() => operation.PostOperationCompleted(noopCallback, null));
+            Assert.Throws<InvalidOperationException>(() => operation.OperationCompleted());
+        }
+
+        [Fact]
+        public static void Cancel()
+        {
+            // Test that cancellation gets passed all the way through PostOperationCompleted(callback, AsyncCompletedEventArgs)
+            var operation = new TestAsyncOperation(op =>
+            {
+                var startTime = DateTime.Now;
+                var timeoutTime = startTime.AddSeconds(k_spinTimeoutSeconds);
+            
+                while (!op.Cancelled)
```

Comment:
Do we really need to spin like this?  Can't we do something such as setting an event from the callback and then just waiting on that event (e.g. a ManualResetEventSlim) with the relevant timeout?